include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -0,0 +1,836 @@
+// -*- C++ -*-
+//===-- parallel_backend_sycl_reduce_then_scan.h ---------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_PARALLEL_BACKEND_SYCL_REDUCE_THEN_SCAN_H
+#define _ONEDPL_PARALLEL_BACKEND_SYCL_REDUCE_THEN_SCAN_H
+
+#include <algorithm>
+#include <cstdint>
+#include <type_traits>
+
+#include "sycl_defs.h"
+#include "parallel_backend_sycl_utils.h"
+#include "execution_sycl_defs.h"
+#include "unseq_backend_sycl.h"
+#include "utils_ranges_sycl.h"
+
+#include "../../utils.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __par_backend_hetero
+{
+
+template <std::uint8_t __sub_group_size, bool __init_present, typename _MaskOp, typename _InitBroadcastId,
+          typename _SubGroup, typename _BinaryOp, typename _ValueType, typename _LazyValueType>
+void
+__exclusive_sub_group_masked_scan(const _SubGroup& __sub_group, _MaskOp __mask_fn, _InitBroadcastId __init_broadcast_id,
```

Comment:
Why subgroup is defined with a template parameter in this and all other helper functions, instead of being just `sycl::sub_group`? It's not even a template class in SYCL.

Reply:
If I recall, this might have something to do with the `sub_group` class existing in a separate extensions namespace in older versions of libsycl. A similar pattern occurs in radix sort where we have to query for the subgroup type: https://github.com/oneapi-src/oneDPL/blob/82aa13f3060e144e7bc2312df47702b8a0ea285e/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h#L400

Higher up in that file, we only use `sycl::sub_group` directly if we know that we are using a version of libsycl of at least 5.7.

Reply:
Can we handle it by defining a type alias in `namespace __dpl_sycl` that depends on the version of libsycl - similarly as we do it for many APIs in `hetero/dpcpp/sylc_defs.h`?

Reply:
Yes, that is a good suggestion. I made this change.